cnf-certification-test/accesscontrol/rbac/roles.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2022 Red Hat, Inc.
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package rbac
+
+import (
+	"strings"
+
+	"github.com/test-network-function/cnf-certification-test/pkg/stringhelper"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+type RoleRule struct {
+	Resource RoleResource
+	Verb     string
+}
+type RoleResource struct {
+	Group, Name string
+}
+
+type CrdResource struct {
+	Group, SingularName, PluralName string
+	ShortNames                      []string
+}
+
+func GetCrdResources(crds []*apiextv1.CustomResourceDefinition) (resourceList []CrdResource) {
+	for _, crd := range crds {
+		var aResource CrdResource
+		aResource.Group = crd.Spec.Group
+		aResource.SingularName = crd.Spec.Names.Singular
+		aResource.PluralName = crd.Spec.Names.Plural
+		aResource.ShortNames = crd.Spec.Names.ShortNames
+		resourceList = append(resourceList, aResource)
+	}
+	return resourceList
+}
+
+func GetAllRules(aRole *rbacv1.Role) (ruleList []RoleRule) {
+	for _, aRule := range aRole.Rules {
+		for _, aGroup := range aRule.APIGroups {
+			for _, aResource := range aRule.Resources {
+				for _, aVerb := range aRule.Verbs {
+					var aRoleRule RoleRule
+					aRoleRule.Resource.Group = aGroup
+					aRoleRule.Resource.Name = aResource
+					aRoleRule.Verb = aVerb
+					ruleList = append(ruleList, aRoleRule)
+				}
+			}
+		}
+	}
+	return ruleList
+}
+
+// Checks the resource name in the role against plural, singular and short names
+func equalResource(aResource CrdResource, aRule RoleRule) bool {
+	// remove subresources to keep only resource name
+	ruleBaseResourceName := strings.Split(aRule.Resource.Name, "/")[0]
+	return aResource.Group == aRule.Resource.Group && (aResource.SingularName == ruleBaseResourceName ||
+		aResource.PluralName == ruleBaseResourceName ||
+		stringhelper.StringInSlice(aResource.ShortNames, ruleBaseResourceName, false))
+}
+
+func FilterRulesNonMatchingResources(ruleList []RoleRule, resourceList []CrdResource) (matching, nonMatching []RoleRule) {
+	for _, aRule := range ruleList {
+		for _, aResource := range resourceList {
+			if equalResource(aResource, aRule) {
+				matching = append(matching, aRule)
```

Comment:
Here, you've confirmed that this role's rule has the crd resource as "target", but shouldn't you also check that its parent role has only one rule? Otherwise, the role could have more rules for other resources, which is what we don't want, as per the second part of the TC description "If an application creates CRDs it must supply a role to access those CRDs **and no other API resources/permission**.".

Reply:
This part only matches the valid rules in the role, e.g. the ones that apply to CRDs under test.  The rules applying to non-CRDs under test in the same role are picked up by:
nonMatching = SliceDifference(ruleList, matching)
in my understanding "If an application creates CRDs it must supply a role to access those CRDs and no other API resources/permission." means "If an application creates CRDs it must supply a role to access those CRDs only and no API resources/permission other than what has been created by the application." It does not limit the number of rules in a role or the number of roles

Reply:
You're right, your code looks good.